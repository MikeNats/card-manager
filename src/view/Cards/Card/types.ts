import { CardModel } from '../../../domain/card/entity'

export type CardProps= {
  cardData: CardModel
  toggleModal:Function
  edit: boolean
}