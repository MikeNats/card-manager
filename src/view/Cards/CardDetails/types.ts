import { CardModel } from '../../../domain/card/entity'

export type CardDetailsProps = { 
    cardData: CardModel,
}