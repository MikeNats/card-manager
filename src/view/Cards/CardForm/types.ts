import { CardModel } from '../../../domain/card/entity'
import {ReactText} from 'react'
export type CardFormProps = { 
  operation: OPERATION,
  cardData: CardModel | CardModelInput,
  newInstance: boolean
}

export enum OPERATION{
  EDIT= 'EDIT',
  CREATE= 'CREATE'
}

export interface CardModelInput{
  cardNumber: string
  id: string;
  type: string;
  name: string;
  expiration: string; 
  cvc: ReactText;
}

