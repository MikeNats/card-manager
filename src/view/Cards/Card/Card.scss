.comp-card{
  padding: 24px;
  background-color: $purple;
  border-radius: 16px;
  margin-bottom: 16px;
  background-image: url('/card-background-shape.svg');
  background-repeat: no-repeat;
  background-position: top right;
  height: 184px;
  width: 311px;
 
  .top{
    display: flex;
    background-repeat: no-repeat;
    background-position: top left;
    flex-direction: row-reverse;
    margin-bottom: 60px;
    :nth-of-type(2){
      margin-right: 24px;
    }
    .info{
      h5{
        color: $grey-50;
        font-size: 10px;
        margin: 0 0 8px 0
      }
      span{
        color: $white;
        font-size: 14px;
        font-weight: bold;
        display: block;
      }
    }
  }
  .bottom{
    position: relative;
    .card-name{
      color: $white;
      font-weight: bold;
      font-size: 14px;
    }
    .card-number{
      color: $grey-50;
      font-weight: bold;
      font-size: 14px;
    }
  }
  .comp-button.toggle-modal{
    height: 22px;
    width: 22px;
    background-image: url('/edit-icon.svg');
    background-repeat: no-repeat;
    background-position: center center;
    border: none;
    outline: none;
    background-color: transparent;
    position: absolute;
    right: 0;
    top: 18px;
    margin:0;
    padding: 0;
    cursor: pointer;
  }
} 
.comp-card.visa{
  .top{
    background-image: url('/visa-logo.svg');
  }

}
.comp-card.mastercard{ 
  .top{ 
    background-image: url('/mastercard-logo.svg');
  }


}