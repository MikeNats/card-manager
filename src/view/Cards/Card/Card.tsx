import React from 'react';
import { CardProps } from './types'
import Button from '../../common/Button/Button'
import { addSpaceEveryFourChars } from '../../../domain/card/cardService'

const Card = ({cardData, toggleModal, edit}: CardProps) => {
  const hideModal = () => toggleModal(true); 
  
  return (<div className={`comp-card ${cardData.type}`}>
    <div className="top">
      <div className="info">
        <h5>CVC</h5>
        <span>{cardData.cvc}</span> 
      </div>
        <div className="info">
          <h5>EXPIRES</h5>
          <span>{cardData.expiration}</span>
        </div>
    </div> 
    <div className="bottom">
      <div><span className="card-name">{cardData.name}</span></div>
      <div><span className="card-number">{addSpaceEveryFourChars(cardData.cardNumber)}</span></div>
      <Button onClickHandler={hideModal} className={`toggle-modal ${edit ? '' : 'hide' }`}/>
    </div>
</div>)
}


export default React.memo(Card)