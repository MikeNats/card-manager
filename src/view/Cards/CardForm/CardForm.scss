.comp-card-form{
    margin-top: 15px;
    .comp-card{
        margin-bottom: 35px;
    }
    fieldset{
        position: relative;
    }
    label{
        font-weight: bold;
        font-size: 14px;
        color: #000;
        margin-bottom: 14px;
    }
    input{
        font-size: 14px;
        color: #000;
        width:100%;
        border: none;
        border-bottom: 1px solid $grey-20;
        padding-bottom: 10px;
        outline: none;
    }
    ::placeholder { 
        color: grey-20;
      } 
    input {
        background-position: 100% 0%;
        background-repeat: no-repeat;
    }
    input.error{
        background-image: url('/form-error.svg');
    }
    .valid input{
        background-image: url('/form-success.svg');
        border-color: $green-30;
        color: $green-30;  
    }
    input.error, .errorMessage{
        color: $red-30;
        border-color: $red-30;

    }
    .errorMessage{
        margin:0;
        font-size: 14px;
        position: absolute;
        bottom: -12px;    
    }
    .errorMessage.hide{
        display: none;
    }
    button:disabled {
        background-color: $grey-20;
    }
    button{
        margin: 0 auto 48px auto;
    }
    button:focus{
        outline: none;
    }
    #delete-card-button{
        background: none;
        color: $grey-20;
        padding:0;
        &:hover{
            color: #000;
        }
    }
}