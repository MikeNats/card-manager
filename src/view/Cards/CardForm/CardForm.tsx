import React, {useEffect, useState} from 'react';
import { isValidCardName, isValidCardNumber, isValidDate, isValidCvc } from '../../../domain/card/cardService'
import Input from '../../common/Input/Input'
import { throttle } from 'lodash'
import {CardFormProps, OPERATION} from './types' 
import Button from '../../common/Button/Button'
import {  createUpdateCard } from './utils'
import { deleteCrad } from '../../../service/Cards'
import { addSpaceEveryFourChars, addSlashEveryTwoChars } from '../../../domain/card/cardService'
  

const CardForm = function({operation, cardData, newInstance}: CardFormProps) {
  const [name, setName] = useState(cardData.name || '')
  const [cardNumber, setCardNumber] = useState(addSpaceEveryFourChars(cardData.cardNumber || ''))
  const [expiration, setExpiration] = useState(cardData.expiration || '')
  const [cvc, setCvc] = useState(cardData.cvc || '')  

  const [invalidname, setInvalidName] = useState(false)
  const [invalidCardNumber, setInvalidCardNumber] = useState(false)
  const [invalidexpiration, setInvalidexpiration] = useState(false)
  const [invalidCvc, setInvalidCvc] = useState(false)
  const [asyncError, setAsyncError] = useState(false)
  const [validName, seValidName] = useState(false)
  const [validCardNumber, setValidCardNumber] = useState(false)
  const [validexpiration, setValidexpiration] = useState(false)
  const [validCvc, setValidCvc] = useState(false)
   
  useEffect(() => {
     setAsyncError(false)
    if(operation === OPERATION.CREATE) {
      setName('') 
      setCardNumber('')
      setExpiration('')
      setCvc('')
    }

    seValidName(false)
    setValidCardNumber(false) 
    setValidexpiration(false)
    setValidCvc(false)
  }, [newInstance, operation])

  const submitForm = (e: React.FormEvent<HTMLFormElement>) =>{
    e.preventDefault();
    validateForm();
    if (!invalidname &&
      !invalidCardNumber &&
      !invalidexpiration &&
      !invalidCvc) {
        createUpdateCard(operation,
          {id: cardData.id, 
          type: cardData.type,
          name, cardNumber, 
          expiration,  
          cvc}).catch(e => {
            setAsyncError(true)})
      }
  }
  const deleteCard = () => 
    deleteCrad(cardData.id)
      .catch(e => setAsyncError(true))

  const onnameChange = throttle(name => 
    setName(name), 200)

  const onCardNumberChange = throttle(inputCardNumber => 
    setCardNumber(addSpaceEveryFourChars(inputCardNumber)), 200)

  const onexpirationChange = throttle(inputexpiration => 
    setExpiration(addSlashEveryTwoChars(inputexpiration)), 200) 

  const onCvcChange = throttle(inputCvc => 
    setCvc(inputCvc), 200)

  const validateForm = () => {
    nameValidation()
    cardNumberValidation()
    expirationValidation()
    cvcValidation()
  }

  const nameValidation = () => {
    if (!isValidCardName(name)) {
      setInvalidName(true)
      seValidName(false)
    } else {
      setInvalidName(false)
      seValidName(true)
    }
  }

  const cardNumberValidation = () => {
    if (!isValidCardNumber(cardNumber)) {
      setInvalidCardNumber(true)
      setValidCardNumber(false)
    } else {
      setInvalidCardNumber(false)
      setValidCardNumber(true)
    }
  }
      

  const expirationValidation = () => {
    if(!isValidDate(expiration)) {
      setInvalidexpiration(true) 
      setValidexpiration(false)
    } else {
      setInvalidexpiration(false)
      setValidexpiration(true)
    }
  }

  const cvcValidation = () =>{
   if(!isValidCvc(cvc)) {
    setInvalidCvc(true)
    setValidCvc(false)
   } else {
    setInvalidCvc(false)
    setValidCvc(true)
   }
  }
   
  return (<section className="comp-card-form">
    { asyncError ? <h3>Oops somthing went wrong...</h3> :
      <>
      <form onSubmit={submitForm}>
          <Input 
              title="Name in card"
              placeholder="Jhon Cabruci"
              type="text"
              id="card-name"
              name="card-name"
              errorMessage="Please fill your name"
              className={validName ? 'valid' : ''}
              onBlur={nameValidation}
              value={name}
              error={invalidname}
              onChangeHandler={onnameChange}/>
          <Input 
            title="Card number"
            type="text"
            placeholder="5532 1234 5545 8014"
            id="card-number"
            name="card-number"
            errorMessage="Please enter a valid card number"
            className={validCardNumber ? 'valid' : ''}
            onBlur={cardNumberValidation}
            maxLength={19}  
            value={cardNumber}  
            error={invalidCardNumber}
            onChangeHandler={onCardNumberChange}/>
          <Input
            title="Expirecy date"
            placeholder="08/21"
            type="text"
            id="expirecy-date"
            name="expirecy-date"
            maxLength={5} 
            errorMessage="Please enter a valid expirecy date"
            className={validexpiration ? 'valid' : ''}
            onBlur={expirationValidation}
            value={expiration}
            error={invalidexpiration}
            onChangeHandler={onexpirationChange}/>
          <Input
            title="CVC (Security code)"
            placeholder="009"
            type="text" 
            id="cvc"
            name="cvc"
            errorMessage="Please enter a valid security code"
            className={validCvc ? 'valid' : ''}
            onBlur={cvcValidation}
            value={`${cvc}`}
            maxLength={3}  
            error={invalidCvc}
            onChangeHandler={onCvcChange}/>

          <Button 
            id="create-update-card-button" 
            disabled= {!!(
              !name &&
              !cardNumber &&
              (!expiration && expiration.split("").length !==5 )  &&
              (!cvc && `${cvc}`.split("").length !==3 ))
            }>Confirm</Button> 
      </form>
     <Button 
        id="delete-card-button"
        disabled= {!cardData.id}
        className={operation === OPERATION.EDIT ? '' : 'hide'} 
        onClickHandler={deleteCard}>Delete Card</Button></>
      }
    </section>
  ) 
} 

export default React.memo(CardForm)