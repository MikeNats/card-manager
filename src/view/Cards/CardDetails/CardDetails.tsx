import React, {useState} from 'react';
import Card from '../Card/Card'
import Modal from '../../common/Modal/Modal'
import CardForm from '../CardForm/CardForm'
import { OPERATION } from '../CardForm/types'
import {CardDetailsProps} from './types'

const CardDetails = function({cardData}:CardDetailsProps) {
  const [isModalVisible, setIsModalVisible] = useState(false)
    return (<div className="comp-card-details">
      <Card
        edit={true}
        toggleModal={setIsModalVisible}
        cardData={cardData}/>
      <Modal 
        title="Edit your card"
        isVisible={isModalVisible}  
        setVisibility={setIsModalVisible}>    
        <Card
          toggleModal={setIsModalVisible}
          edit={false}
          cardData={cardData}/> 
        <CardForm 
          newInstance={isModalVisible}
          operation={OPERATION.EDIT}
          cardData={cardData}/>
      </Modal>
  </div>)
} 
 
export default React.memo(CardDetails)