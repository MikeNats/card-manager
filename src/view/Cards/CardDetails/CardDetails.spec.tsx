import React from 'react'
import CardDetails from './CardDetails'
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { CardModel } from '../../../domain/card/entity'
import Card from '../Card/Card'
import Modal from '../../common/Modal/Modal'
import CardForm from '../CardForm/CardForm'
Enzyme.configure({ adapter: new Adapter() })
const mockCard: CardModel ={
  id: "1231234234",
  type: "Mastercard",
  name: "Jhon Done",
  cardNumber: 1231234234,
  expiration: "10/24",
  cvc:432
};
describe("CearDetails", () => {

  it('renders as expected', () => {
    const component = shallow(<CardDetails
      cardData={mockCard}
      />)

    expect(component).toMatchSnapshot();
  })

  it('should render a Card', () => {	
    const component = shallow(<CardDetails
      cardData={mockCard}
    />)

    expect(component.childAt(0).type()).toEqual(Card);
  });

  it('should render a Modal', () => {	
    const component = shallow(<CardDetails
      cardData={mockCard}
    />)

    expect(component.childAt(1).type()).toEqual(Modal);
  });

  it('should Modal have as children: Card and CardForm', () => {	
    const component = shallow(<CardDetails
      cardData={mockCard}
    />)

    expect(component.childAt(1).find(Modal)).toBeTruthy();
    expect(component.childAt(1).find(CardForm)).toBeTruthy();
  });
  
});
