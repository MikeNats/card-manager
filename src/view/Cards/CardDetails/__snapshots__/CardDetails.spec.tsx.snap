// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CearDetails renders as expected 1`] = `ShallowWrapper {}`;
